File: 프로그래머스/lv2/42888. 오픈채팅방/오픈채팅방.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

class Solution {
    public String[] solution(String[] record) {
        HashMap<String,String> map = new HashMap<>();
        List<String> list= new ArrayList<>();

        for(int i = record.length - 1; i >= 0; i--){
            String[] arr = record[i].split(" ");
            if(!map.containsKey(arr[1]) && !arr[0].equals("Leave")){
                map.put(arr[1], arr[2]);
            }
        }

        for(String temp : record){
            String[] arr = temp.split(" ");
            if(arr[0].equals("Enter")){
                list.add(map.get(arr[1]) + "님이 들어왔습니다.");
            }
            else if(arr[0].equals("Leave")){
                list.add(map.get(arr[1]) + "님이 나갔습니다.");
            }
            else{
                continue;
            }
            
        }
        return list.toArray(new String[list.size()]);
    }
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    map = {}
    answer = []
    
    for i in record:
        parts = i.split()
        if len(parts) > 2:
            map[parts[1]] = parts[2]
       
    for j in record:
        parts = j.split()
        nickname = map.get(parts[1])
        if parts[0] == "Enter":
            answer.append(nickname+"님이 들어왔습니다.")
        elif parts[0] == "Leave":
            answer.append(nickname+"님이 나갔습니다.")
    
    return answer